community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,109 +19,176 @@
  */
 package org.neo4j.concurrent;
 
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.LockSupport;
 
 /**
  * Turns multi-threaded unary work into single-threaded stack work.
  * <p>
- *     The technique used here is inspired in part both by the Flat Combining
- *     concept from Hendler, Incze, Shavit &amp; Tzafrir, and in part by the
- *     wait-free linked queue design by Vyukov.
+ * The technique used here is inspired in part both by the Flat Combining
+ * concept from Hendler, Incze, Shavit &amp; Tzafrir, and in part by the
+ * wait-free linked queue design by Vyukov.
  * </p>
  * <p>
- *     In a sense, this turns many small, presumably concurrent, pieces of work
- *     into fewer, larger batches of work, that is then applied to the material
- *     under synchronisation.
+ * In a sense, this turns many small, presumably concurrent, pieces of work
+ * into fewer, larger batches of work, that is then applied to the material
+ * under synchronisation.
  * </p>
  * <p>
- *     Obviously this only makes sense for work that a) can be combined, and b)
- *     where the performance improvements from batching effects is large enough
- *     to overcome the overhead of collecting and batching up the work units.
+ * Obviously this only makes sense for work that a) can be combined, and b)
+ * where the performance improvements from batching effects is large enough
+ * to overcome the overhead of collecting and batching up the work units.
  * </p>
+ *
  * @see Work
  */
 public class WorkSync<Material, W extends Work<Material,W>>
 {
     private final Material material;
     private final AtomicReference<WorkUnit<Material,W>> stack;
     private final WorkUnit<Material,W> stackEnd;
```

Comment:
It can be made static at the cost of some unchecked casts to get around the generics.